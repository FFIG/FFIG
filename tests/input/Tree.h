#include <memory>
#include <random>

#ifdef __clang__
#define C_API __attribute__((annotate("GENERATE_C_API")))
#else
#define C_API
#endif

static std::mt19937 mt;
static std::uniform_int_distribution<int> d(1,10);
static auto gen = []{return d(mt);};

class C_API Tree
{
  int data_;
  std::unique_ptr<Tree> left_;
  std::unique_ptr<Tree> right_;

  public:

  Tree(int levels=0)
  {
    data_ = gen();
    if ( levels <= 0 ) return;
    left_ = std::make_unique<Tree>(levels-1);
    right_ = std::make_unique<Tree>(levels-1);
  }

  const Tree* left_subtree() const
  {
    return left_.get();
  }

  const Tree* right_subtree() const
  {
    return right_.get();
  }

  void erase_data() 
  {
    data_ = 0;
  }

  int data() const
  {
    return data_;
  }
  
  void set_data(int x)
  {
    data_ = x;
  }

};

