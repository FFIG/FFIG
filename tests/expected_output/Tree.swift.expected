// This code was generated by FFIG <http://ffig.org>.
// Manual edits will be lost.
class Tree {

  let obj_ : OpaquePointer

  init(obj:OpaquePointer)
  {
    obj_ = obj
  }

  deinit
  {
    Tree_Tree_dispose(obj_)
  }

  func left_subtree() -> CInt
  {
    let rv = Tree_Tree_left_subtree_noexcept(obj_)
    return rv
  }

  func right_subtree() -> CInt
  {
    let rv = Tree_Tree_right_subtree_noexcept(obj_)
    return rv
  }

  func data() -> CInt
  {
    let rv = Tree_Tree_data_noexcept(obj_)
    return rv
  }

  func set_data(x:CInt) -> CInt
  {
    let rv = Tree_Tree_set_data_noexcept(obj_, x)
    return rv
  }

}
